tlatools/org.lamport.tlatools/src/util/SimpleFilenameToStream.java

```diff
@@ -136,112 +128,96 @@ private String[] getLibraryPaths(final String installationBasePath, String[] lib
-   *  the File.  The first fully-qualified name that refers to a file that actually
-   *  exists is the File returned; but if none exists, the last one tried is returned
+   *  From name, try to find the actual file intended.  If the file is not a normal
+   *  file (for instance, if it is an entry in a JAR file), then copy it to the
+   *  filesystem and return the copied file.
+   *
+   *  <p>The name is treated as a filename, not a TLA+ module name; if you are looking
+   *  for the Integers module, pass <code>"Integers.tla"</code>.
+   *
+   *  <p>The first fully-qualified name that refers to a file that actually
+   *  exists is the File returned; but if none exists, a non-existent file is returned
    *  anyway.  Hence, the File object returned does not necessarily represent a file
    *  that actually exists in the file system.
    *
-   *  @param  module name, used as basis of path name to the file that should contain it
+   *  @param name the name of the file to look for
    */
-  private final File locate(String name)
+  private final TLAFile locate(String name)
   {
+    Optional<TLAFile> userModule = locateUsingResolver(userFileResolver, name, false);
+    if (userModule.isPresent()) {
+        return userModule.get();
+    }
 
-    String prefix      = "";                // directory to be prepended to file name before lookup
-
-    File   sourceFile  = null;              // File object from fully-qualified name of file found by
-                                            //   searching "libraryPath" for "name", as is done for PATHs and CLASSPATHs
+    Optional<TLAFile> libraryModule = locateUsingResolver(standardLibraryResolver, name, true);
+    if (libraryModule.isPresent()) {
+        return libraryModule.get();
+    }
 
-    // The TLA standard modules are presumed to be in a directory tla2sany/StandardModules
-    // i.e. a toplevel subdirectory of the main installation directory for SANY
-    // We append the StandardModules director to the library path, so the user does
-    // not have to do anything special and the standard modules are automatically accessible
+    // Small bit of awkwardness here: to preserve the historical FilenameToStream API, this method must return a
+    // nonexistent file if the requested one could not be located.  The file we return will be used in error messages
+    // about the missing file, so we should return one that looks plausible, but we can't return one that MIGHT exist,
+    // because the result will be read if it does!  The approach below correctly synthesizes a non-existent
+    // believable-looking path, but relies on the fact that ToolIO.getUserDir() is in the search paths used by
+    // userFileResolver---this file does not exist because it has certainly already been searched.  In the future, we
+    // might improve this pattern by changing the FilenameToStream contract to say that it throws an exception or
```

Comment:
This brings back unpleasant memories because the Toolbox suffered and suffers from this awkwardness.  To make matters worse, the Toolbox additionally accesses files through Eclipse's FS abstraction.  Instead of refactoring `FilenameToStream`, it would likely be better to deprecate it and create a new API in tla2tools.

Reply:
Agreed. I don't know exactly what that should look like, but I can at least amend this comment to include that idea.